### PR TITLE
feat(server): use APFS clones to save worktree disk space

### PR DESCRIPTION
**Pull request by saphid**
## What Changed

Uses APFS copy-on-write files when creating eligible clean, same-commit worktrees on macOS. Eligible tracked regular files share disk blocks until either copy is edited; Git verifies their content and repairs differences. Files requiring checkout conversions stay with Git, and cloned files receive normal checkout permissions. Git still materializes tracked symlinks and initializes submodules. Set `"worktreeCloneFiles": false` in `t3.json` to opt out.

An explicit `worktreeCloneDependencies` option also seeds ignored `node_modules` directories only after the effective setup script is selected, just before it runs. Manual creation and disabled setup overrides do not seed dependencies. Old executable shims and known caches are removed from the seed; non-portable symlinks cause the seed to be discarded. Existing destination installs are preserved. This repository opts in.

Unsupported filesystems, immutable APFS sources, and ineligible repositories retain ordinary Git/setup behavior. Conditional Git configuration uses ordinary checkout so destination-only filters cannot leave stale clone contents. No hardlinks are used. The platform operation is separate from the shared worktree logic so Linux and Windows support can follow in separate dependent PRs. The implementation is server-side and covers all clients and providers, including remote environments.

## Why

The goal is lower worktree disk usage with independent edits. Cloning is not presented as a checkout speed improvement: Git verification adds overhead, which is an intentional trade-off for shared disk blocks. Dependency setup still reconciles packages and regenerates executable shims.

## Verification

On a clean macOS/APFS checkout based on upstream `93e04160a0`:

- `vp test run apps/server/src/vcs/WorktreeClone.test.ts apps/server/src/vcs/GitVcsDriverCore.test.ts packages/contracts/src/t3ProjectFile.test.ts packages/shared/src/t3ProjectFile.test.ts apps/server/src/project/ProjectSetupScriptRunner.test.ts`: **139 passed**, with global/system Git configuration disabled for isolated fixtures.
- Targeted lint and `tsc --noEmit --project apps/server/tsconfig.json`: passed.
- Native tests cover copy-on-write/edit isolation, small-file defaults without project configuration, opt-out, source edits/ref movement during creation, clone failure recovery, Git hooks (including configured `core.hooksPath`)/filters/submodules, LF/CRLF and ident conversion, source permission drift, immutable file fallback, relative destination paths, and effective setup-script selection.
- Integrated disposable pnpm fixture: source install → T3 tracked worktree creation → dependency setup seeding → offline install → execute the rebuilt package binary. It resolved inside the target worktree and the source executable stayed intact.
- Disposable isolated APFS volume: a 64 MiB file clone consumed **0 additional allocated data bytes**, versus **64 MiB** for an ordinary streamed copy. The clone had a distinct inode, and editing it left the source unchanged. The image was detached and removed afterward.
- A focused macOS GitHub Actions job verifies the native path independently of the broader Linux CI suite.

This is a disk-space feature, not a claimed checkout speedup. Dependency-install speedup was not benchmarked.

Fresh independent review was attempted directly with `devin -p --model swe-2-max --permission-mode auto --respect-workspace-trust false --prompt-file …`. The final attempt returned no output within 180 seconds and was terminated (actual exit `-15`). No responding reviewer model or verdict is claimed.

## Checklist

- [x] This PR is focused on worktree filesystem copies
- [x] I explained what changed and why
- [x] Focused tests and native macOS verification completed

Implemented with GPT-6 using the Codex harness.

## Summary by CodeRabbit

* **New Features**
  * macOS worktrees can use space-efficient APFS copies for eligible tracked files, with Git fallback when unsupported.
  * Dependencies can be copied into new worktrees when enabled and setup scripts are configured, reducing installation time.
  * Added project settings for controlling file and dependency cloning.
  * Worktree setup now reconciles copied dependencies before running the setup command.
  * Worktree creation preserves executable permissions and supports files with checkout-conversion rules.

* **Documentation**
  * Added guidance for configuring macOS worktree copies, dependency seeding, opt-out settings, and fallback behavior.

### CODE REVIEW

File: apps/server/src/vcs/WorktreeClone.ts [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+        cwd,
+        [
+          "config",
+          "--get-regexp",
+          "^(filter\\.|core\\.sparsecheckout$|extensions\\.worktreeconfig$)",
+        ],
+        true,
+      );
+      if (config.exitCode !== 1) return null;
+      const hook = (yield* git(cwd, [
```
[the remark this answers is not in the document]
Comment (saphid) [reply apps/server/src/vcs/WorktreeClone.ts:62]: Verified this against Git and added a real-driver regression test: `git rev-parse --git-path hooks/post-checkout` already resolves `core.hooksPath`. With an executable hook in a custom configured directory, prepare returns null, ordinary worktree creation runs the hook, and its marker appears in the target. The new regression passes. No production behavior change is needed for this finding.

File: apps/server/src/vcs/WorktreeClone.ts [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+const timeoutMs = 300_000;
+const maxOutputBytes = 16 * 1024 * 1024;
+const decodeProjectFile = Schema.decodeUnknownEffect(Schema.fromJsonString(T3ProjectFile));
+
+/** Seeds an ordinary Git worktree; Git still owns its index and final contents. */
+export const makeWorktreeClone = Effect.fn("makeWorktreeClone")(function* (
+  execute: GitVcsDriver["Service"]["execute"],
+) {
+  const fs = yield* FileSystem.FileSystem;
```
[the remark this answers is not in the document]
Comment (saphid) [reply apps/server/src/vcs/WorktreeClone.ts:20]: This is the constructing driver’s local execution callback, not an acquired GitVcsDriver service instance. It is defined inside makeGitVcsDriverCore before the final service is assembled. The callback keeps clone operations on the owner’s tracing, metrics, timeout and process policy; acquiring GitVcsDriver here would introduce a construction cycle. This fits the documented deliberate-callback-strategy exception. Added a comment explaining the boundary; extracting the existing executor would be a separate refactor.

File: apps/server/src/vcs/FileClone.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      return yield* new GitCommandError({
+        operation: "FileClone.copy",
+        command: "/bin/cp",
+        cwd: destination,
+        detail: "Immutable source files require ordinary checkout",
+      });
+    }
+    const code = yield* spawner
+      .exitCode(
+        ChildProcess.make("/bin/cp", ["-c", "-p", "-R", "-P", ...sources, destination], {
```
[the remark this answers is not in the document]
Comment (saphid) [reply apps/server/src/vcs/FileClone.ts:45]: I reproduced this on APFS with an owner deny-write ACL. Ordinary cp -p retained that ACL, but the actual cp -c -p path did not: ls -le showed no ACL on the clone, and writing the clone succeeded while the source retained its ACL. cp -c without -p behaved the same. This finding applies to an ordinary metadata-preserving copy, not the forced APFS clone path here; no production change is justified by the reproduction.

File: apps/server/src/vcs/WorktreeClone.test.ts [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+        yield* Effect.addFinalizer(() =>
+          spawner
+            .exitCode(
+              ChildProcess.make("/usr/bin/chflags", ["nouchg", source, target], {
+                stderr: "ignore",
+              }),
+            )
+            .pipe(Effect.ignore),
+        );
+        yield* spawner.exitCode(ChildProcess.make("/usr/bin/chflags", ["uchg", source]));
```
[the remark this answers is not in the document]
Comment (saphid) [reply apps/server/src/vcs/WorktreeClone.test.ts]: Fixed in 269c879511: the test now asserts chflags exits zero before exercising worktree creation. Focused clone suite passes all 20 tests, including real immutable-file fallback.

**Review by juliusmarminge — CHANGES_REQUESTED** [review]
Thanks for the careful work here — the edge-case handling (immutable flags, `core.hooksPath`, conditional filters, the race repairs) is thorough and the bot threads were all answered with real reproductions. I still want to request changes, because after tracing this against real machines the default-on part almost never activates, and the part that could matter is opt-in and unmeasured. I also compared the approach with [anomalyco/rift](https://github.com/anomalyco/rift), which attacks the same problem from the other end; notes on that at the bottom.

## Tracked-file cloning is effectively inert on real repositories

`prepare` returns `null` whenever any of these hold. I checked each against an actual dev machine rather than the fixtures:

1. **Any `filter.*` in git config, at any level.** git-lfs writes `filter.lfs.*` into system or global config on install. The PR's own CI job has to set `GIT_CONFIG_GLOBAL=/dev/null` for exactly this reason — hosted runners have it, and so do most developer machines with lfs installed.
2. **`extensions.worktreeconfig`.** Set by `git sparse-checkout` and various worktree tooling; the maintainers' main t3code checkout has it.
3. **Any file with a `text` or `eol` attribute.** `text=auto` is treated as a conversion and excluded. t3code's `.gitattributes` is `* text=auto eol=lf`, which is also GitHub's recommended template. Running the exact `check-attr` query `prepare` uses over t3code's tree: all 23,076 tracked files come back `auto`, so `cloneFiles` is empty and **the clone path never executes on this repository, on any platform**. As far as I can tell none of the verification runs exercised it on t3code itself.
4. **`core.autocrlf` set to anything but false.** Git for Windows' installer writes `core.autocrlf=true` to system config by default, so on Windows this is off for nearly everyone regardless of ReFS (relevant to #12632).

Even where it does run, the tracked tree is small (t3code: ~35 MB tracked vs ~3 GB `node_modules`), and the PR documents that checkout gets *slower* (`update-index --refresh` re-hashes every file, plus a per-file `readLink`+`stat`+`chmod` loop, plus `reset --hard`). Default-on, documented-slower, for tens of megabytes is the wrong trade.

## Dependency seeding is the feature, but it's a speed feature for pnpm, not a disk one

pnpm already hardlinks/reflinks every package from the content-addressable store, so a fresh install in a worktree costs almost no extra disk. Bun clones on macOS by default. The disk argument only holds for npm/yarn-classic. For pnpm the only possible win is install time, and the description says that was not benchmarked — yet `t3.json` opts this repo in. I'd want a before/after `vp i` number on t3code before accepting that line.

Also: `node_modules/.pnpm-workspace-state-v1.json` contains the source worktree's absolute path (38 occurrences here). The seed copies it verbatim and relies on the install to reconcile. Whether pnpm treats a mismatched state file as "reinstall everything" (making the seed pure overhead) is unverified for a monorepo of this size.

## Blocking: unsupported filesystems pay the full cost every time

`supported` is a platform check, not a filesystem check, and nothing caches an "unsupported" answer. This is mostly a #12634/#12632 problem, but the shape is set here:

- On Linux/ext4 (most Linux developers), `prepare` still produces a plan, `cp --reflink=always -R` fails, and `reset --hard` finishes the checkout. Fine for tracked files, but for dependency seeding I verified that GNU `cp --reflink=always -R` **walks the entire source tree** before failing (400 errors for a 400-file tree, exit 1). With `worktreeCloneDependencies: true`, every worktree creation walks ~250k `node_modules` entries into a staging dir, fails, and deletes it.
- On Windows/NTFS the equivalent is spawning PowerShell and `Add-Type`-compiling the C# helper on every attempt before the volume check throws.
- When the clone attempt fails after `--no-checkout`, the real checkout happens inside `reset --hard` with no progress wiring, so the checkout progress bar in the UI sits still until 100%. That's a regression from today's behavior on exactly the fallback path.

Suggested fix: probe once per (source, destination) with a real clone — Node's `fs.copyFile(src, dst, COPYFILE_FICLONE_FORCE)` on a single file — and skip everything on failure. The Linux test already uses precisely this as its oracle; production should too. On macOS and Linux that call could replace the `find`/`cp` subprocesses entirely (it's `clonefile`/`FICLONE` under the hood).

## Other

- `warmDependencies` runs before the setup terminal is spawned, so cloning and then walking every staged entry (`readDirectory`/`readLink`/`stat` per entry for the portability check) gives the user no feedback. Unmeasured, but not free at 250k entries.
- `macos-worktree-clone.yml` triggers a macOS runner on any PR touching `GitVcsDriverCore.ts`, a ~3k-line file that changes constantly. That's a recurring paid-minutes decision for the maintainers to make explicitly.
- #12634 and #12632 each vendor this entire diff and modify `FileClone.ts` in incompatible ways, so whichever lands second needs a manual merge. If this proceeds it should be one stack (or one PR with a platform table in `FileClone`).

## Comparison with rift

rift snapshots the whole directory (including `.git`) with `clonefile` / `FICLONE` / btrfs subvolumes, skips `node_modules`, `target`, `.venv`, `dist` etc. by default, and delegates dependency setup to a `postcreate` hook. It probes reflink support once at `rift init` and refuses to register unsupported directories. It produces an independent detached-HEAD copy rather than a linked worktree, so it's not a drop-in for T3 Code (checkpointing, worktree discovery and cleanup all assume `git worktree`), and it's experimental with no Windows support. But it validates the diagnosis: the win is in the regenerable artifacts and in *not* re-verifying through Git, and support should be probed once with a real syscall rather than re-discovered per attempt. The PRs currently invert both.

## What I'd accept

A much smaller PR that keeps the linked-worktree architecture and:

- drops tracked-file cloning, or makes it opt-in and uses `git ls-files --eol` to include `text=auto` files whose index/worktree line endings already match, instead of excluding on attribute presence;
- makes opt-in dependency seeding the feature, probes filesystem support once natively, and surfaces progress through the setup terminal;
- comes with a measured install-time before/after on t3code for pnpm.

Roughly "rift's copy policy inside `git worktree add`". Happy to re-review a reshaped version. I'm converting #12634 and #12632 to drafts until the foundation settles.

<sub>Audit verification: read the full diff at 8d92b9f7 against `93e04160`; ran the Linux-runnable suites from the #12634 head on Linux/ext4 (29 passed, 1 skipped); reproduced the `cp --reflink=always` full-tree walk locally; did not run the macOS/Windows native paths and relied on the PR's CI for those.</sub>
